File: packages/cli/src/templates/_shared/lib/base/hook_utils.py
"""Common utilities for hook scripts.

Provides standardized boilerplate for:
- Path setup for imports
- JSON parsing from stdin
- Hook payload validation
- Error handling decorators
"""

import json
import os
import sys
from datetime import datetime, timezone
from functools import wraps
from pathlib import Path
from typing import Any, Callable, Dict, Optional, TypeVar

from .logger import log_hook_error, hook_log, log_debug, log_info, log_warn, log_error, log_diagnostic, set_context_path, set_session_id


# Context window baseline: tokens not visible in hook data
# (system prompt, tools, MCP tokens)
# See: https://github.com/anthropics/claude-code/issues/13783
CONTEXT_BASELINE_TOKENS = 22_600
DEFAULT_CONTEXT_WINDOW_SIZE = 200_000


def parse_context_window(hook_input: dict) -> tuple:
    """Parse context window from hook input.

    Returns (tokens_used, max_tokens) or (None, None).
    tokens_used includes baseline offset for system prompt/tools.
    """
    context_window = hook_input.get("context_window")
    if not context_window:
        return None, None
    current_usage = context_window.get("current_usage")
    if not current_usage:
        return None, None
    cache_read = current_usage.get("cache_read_input_tokens", 0) or 0
    input_tokens = current_usage.get("input_tokens", 0) or 0
    cache_creation = current_usage.get("cache_creation_input_tokens", 0) or 0
    output_tokens = current_usage.get("output_tokens", 0) or 0
    content_tokens = cache_read + input_tokens + cache_creation + output_tokens
    tokens_used = content_tokens + CONTEXT_BASELINE_TOKENS
    max_tokens = context_window.get("context_window_size") or DEFAULT_CONTEXT_WINDOW_SIZE
    return tokens_used, max_tokens


def get_context_percent_remaining(hook_input: dict) -> tuple:
    """Get context percentage remaining with context.json fallback.

    Tries two sources in order:
    1. Hook input context_window data (most accurate, real-time)
    2. context.json remaining_percentage (written by status_line.py)

    Returns:
        (percent_remaining, tokens_used, max_tokens) where tokens_used and
        max_tokens may be None if data came from context.json fallback.
        Returns (None, None, None) if no data available from either source.
    """
    # Source 1: Hook input (most accurate)
    tokens_used, max_tokens = parse_context_window(hook_input)
    if tokens_used is not None and max_tokens is not None and max_tokens > 0:
        remaining = max_tokens - tokens_used
        percent_remaining = max(0, min(100, int((remaining / max_tokens) * 100)))
        return percent_remaining, tokens_used, max_tokens

    # Source 2: context.json fallback (written by status_line.py)
    try:
        from .utils import project_dir
        from ..context.context_store import get_context_by_session_id

        session_id = hook_input.get("session_id")
        if session_id:
            project_root = project_dir(hook_input)
            context = get_context_by_session_id(session_id, project_root)
            if context and context.last_session:
                pct = context.last_session.get("context_remaining_pct")
                if pct is not None:
                    return pct, None, None
    except Exception:
        pass  # Fallback failed — degrade gracefully

    return None, None, None


# Type variable for generic decorators
F = TypeVar('F', bound=Callable[..., Any])

# Event metadata stash — populated by load_hook_input(), read by run_hook()
_last_hook_event: Optional[str] = None
_last_tool_name: Optional[str] = None
_last_session_id: Optional[str] = None


def load_hook_input() -> Optional[Dict[str, Any]]:
    """
    Load and parse JSON from stdin.

    Returns:
        Parsed JSON dict, or None if stdin is empty or invalid JSON
    """
    global _last_hook_event, _last_tool_name, _last_session_id
    try:
        input_data = sys.stdin.read().strip()
        if not input_data:
            return None
        result = json.loads(input_data)
        if isinstance(result, dict):
            _last_hook_event = result.get("hook_event_name")
            _last_tool_name = result.get("tool_name")
            _last_session_id = result.get("session_id")
        return result
    except json.JSONDecodeError:
        return None


def validate_hook_event(
    payload: Dict[str, Any],
    expected_event: str,
    expected_tool: Optional[str] = None
) -> bool:
    """
    Validate hook event type and optional tool name.

    Args:
        payload: Hook payload from stdin
        expected_event: Expected hook_event_name (e.g., "PostToolUse", "PreToolUse")
        expected_tool: Optional expected tool_name (e.g., "TaskCreate")

    Returns:
        True if payload matches expected event/tool, False otherwise
    """
    if payload.get("hook_event_name") != expected_event:
        return False
    if expected_tool and payload.get("tool_name") != expected_tool:
        return False
    return True


def get_tool_input(payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """
    Extract and validate tool_input from payload.

    Args:
        payload: Hook payload from stdin

    Returns:
        tool_input dict, or None if missing/invalid
    """
    tool_input = payload.get("tool_input", {})
    return tool_input if isinstance(tool_input, dict) else None


def check_skip_persistence(payload: Dict[str, Any], hook_name: str = "hook") -> bool:
    """
    Check if persistence should be skipped based on metadata flags.

    Args:
        payload: Hook payload from stdin
        hook_name: Name of hook for logging

    Returns:
        True if skip_persistence flag is set, False otherwise
    """
    tool_input = get_tool_input(payload)
    if not tool_input:
        return False

    metadata = tool_input.get("metadata", {})
    if isinstance(metadata, dict) and metadata.get("skip_persistence"):
        log_debug(hook_name, "Skipping persistence (skip_persistence flag set)")
        return True
    return False


def safe_hook_main(hook_name: str) -> Callable[[F], F]:
    """
    Decorator for hook main functions with standard error handling.

    Catches exceptions, logs them to stderr, and returns 0 (non-blocking).

    Args:
        hook_name: Name of hook for error messages

    Returns:
        Decorator function

    Example:
        @safe_hook_main("my_hook")
        def main() -> int:
            # ... hook logic ...
            return 0
    """
    def decorator(func: F) -> F:
        @wraps(func)
        def wrapper(*args, **kwargs):
            try:
                return func(*args, **kwargs)
            except json.JSONDecodeError as e:
                import traceback
                tb = traceback.format_exc()
                log_hook_error(hook_name, e, traceback_str=tb)
                log_error(hook_name, f"JSON decode error: {e}")
                return 0
            except Exception as e:
                import traceback
                tb = traceback.format_exc()
                log_hook_error(hook_name, e, traceback_str=tb)
                log_error(hook_name, f"Unexpected error: {e}", traceback_str=tb)
                return 0
        return wrapper  # type: ignore
    return decorator


def emit_context(additional_context: str, ensure_ascii: bool = False) -> None:
    """Emit hookSpecificOutput with additionalContext to stdout.

    Args:
        additional_context: Context string to inject into Claude's context
        ensure_ascii: If True, escape non-ASCII characters in JSON output
    """
    out = {
        "hookSpecificOutput": {
            "additionalContext": additional_context,
        }
    }
    print(json.dumps(out, ensure_ascii=ensure_ascii))


def emit_context_and_block(
    additional_context: str,
    reason: str,
    ensure_ascii: bool = True,
) -> None:
    """Emit hookSpecificOutput that denies the tool call with context and reason.

    Args:
        additional_context: Context string to inject into Claude's context
        reason: Reason shown to Claude for why the tool call was denied
        ensure_ascii: If True, escape non-ASCII characters in JSON output
    """
    out = {
        "hookSpecificOutput": {
            "additionalContext": additional_context,
            "permissionDecision": "deny",
            "permissionDecisionReason": reason,
        }
    }
    print(json.dumps(out, ensure_ascii=ensure_ascii))


def _detect_template(script_path: str = "") -> str:
    """Auto-detect template origin from the hook script path.

    Returns "shared", a template name (e.g., "cc-native"), or "unknown".
    """
    import re
    path = (script_path or (sys.argv[0] if sys.argv else "")).replace("\\", "/")
    if "/_shared/hooks/" in path or path.startswith("_shared/hooks/"):
        return "shared"
    match = re.search(r'_([a-z][a-z0-9-]*)/hooks/', path)
    if match:
        return match.group(1)  # e.g., "cc-native"
    return "unknown"


def run_hook(main_func: Callable[[], int], hook_name: str = "unknown") -> None:
    """
    Standard hook entry point wrapper with lifecycle logging.

    Logs HOOK_START before calling main, HOOK_END after completion.
    Catches unhandled exceptions and logs them before exiting cleanly.

    Args:
        main_func: Hook main function that returns exit code
        hook_name: Name of the hook for error logging

    Example:
        if __name__ == "__main__":
            run_hook(main, "my_hook")
    """
    import time
    start_time = time.monotonic()
    template = _detect_template()
    event = _last_hook_event or "unknown"
    tool = _last_tool_name

    # Wire session_id into logger so all log entries carry it
    if _last_session_id:
        set_session_id(_last_session_id)

    # HOOK_START
    start_data: Dict[str, Any] = {"lifecycle": "start", "template": template, "event": event}
    if tool:
        start_data["tool"] = tool
    log_info(hook_name, "HOOK_START", data=start_data)

    exit_code = 0
    status = "success"
    error_info = None

    try:
        result = main_func()
        exit_code = result if isinstance(result, int) else 0
        status = "blocked" if exit_code != 0 else "success"
    except SystemExit as e:
        exit_code = e.code if isinstance(e.code, int) else (1 if e.code else 0)
        status = "blocked" if exit_code != 0 else "success"
    except Exception as e:
        import traceback
        exit_code = 0  # Non-blocking
        status = "error"
        error_info = (e, traceback.format_exc())

    # HOOK_END
    duration_ms = round((time.monotonic() - start_time) * 1000, 1)
    end_data: Dict[str, Any] = {
        "lifecycle": "end", "status": status,
        "duration_ms": duration_ms, "exit_code": exit_code,
        "template": template,
    }
    end_event = _last_hook_event or event  # Re-read after main() populated it
    end_tool = _last_tool_name or tool
    end_data["event"] = end_event
    if end_tool:
        end_data["tool"] = end_tool
    if error_info:
        e, tb = error_info
        end_data["error_type"] = type(e).__name__
        log_hook_error(hook_name, e, traceback_str=tb)
        log_error(hook_name, f"HOOK_END: {e}", data=end_data, traceback_str=tb)
    elif status == "blocked":
        log_warn(hook_name, "HOOK_END", data=end_data)
    else:
        log_info(hook_name, "HOOK_END", data=end_data)

    raise SystemExit(exit_code)

File: packages/cli/src/templates/_shared/lib-ts/hooks/hook-utils.ts
/**
 * Common utilities for hook scripts.
 * Standardized boilerplate for JSON parsing, validation, error handling.
 * See SPEC.md §5
 */

import * as fs from "node:fs";

import { getContextBySessionId } from "../context/context-store.js";
import { getProjectRoot } from "../runtime/constants.js";
import { logDebug,  logWarn,     hookLog, setSessionId,  getContextPath as _getContextPath } from "../runtime/logger.js";
import type { HookInput, HookOutput, PermissionRequestOutput } from "../types.js";

// Re-export logger functions for convenience (matches Python hook_utils re-exports)


// Context window baseline: tokens not visible in hook data §5.9
export const CONTEXT_BASELINE_TOKENS = 22_600;
export const DEFAULT_CONTEXT_WINDOW_SIZE = 200_000;

// Event metadata stash — populated by loadHookInput(), read by runHook()
let _lastHookEvent: string | null = null;
let _lastToolName: string | null = null;
let _cachedHookName: string | null = null;

// Pre-fetched input stash
let _prefetchedInput: Record<string, unknown> | null = null;

function readStringField(value: Record<string, unknown>, key: string): null | string {
  const field = value[key];
  return typeof field === "string" ? field : null;
}

/**
 * Load and parse JSON from stdin (or return prefetched input if set).
 * Returns null if stdin is empty or invalid JSON.
 * See SPEC.md §5.1
 */
export function loadHookInput(): HookInput | null {
  if (_prefetchedInput !== null) {
    const result = _prefetchedInput;
    _prefetchedInput = null; // consume once
    if (result && typeof result === "object") {
      _lastHookEvent = readStringField(result, "hook_event_name");
      _lastToolName = readStringField(result, "tool_name");
    }
    return result as unknown as HookInput;
  }

  try {
    // Read entire stdin using fd 0 (cross-platform, works on Windows)
    const inputData = fs.readFileSync(0, "utf8").trim();
    if (!inputData) return null;

    const result = JSON.parse(inputData) as Record<string, unknown>;
    if (result && typeof result === "object") {
      _lastHookEvent = readStringField(result, "hook_event_name");
      _lastToolName = readStringField(result, "tool_name");
    }
    return result as unknown as HookInput;
  } catch {
    return null;
  }
}

/**
 * Validate hook event type and optional tool name.
 * See SPEC.md §5.2
 */
export function validateHookEvent(
  payload: HookInput,
  expectedEvent: string,
  expectedTool?: string,
): boolean {
  if (payload.hook_event_name !== expectedEvent) return false;
  if (expectedTool && payload.tool_name !== expectedTool) return false;
  return true;
}

/**
 * Extract and validate tool_input from payload.
 * See SPEC.md §5.3
 */
export function getToolInput(
  payload: HookInput,
): Record<string, unknown> | null {
  const toolInput = payload.tool_input;
  return toolInput && typeof toolInput === "object" ? toolInput : null;
}

/**
 * Check if persistence should be skipped based on metadata flags.
 * See SPEC.md §5.4
 */
export function checkSkipPersistence(
  payload: HookInput,
  hookName = "hook",
): boolean {
  const toolInput = getToolInput(payload);
  if (!toolInput) return false;

  const {metadata} = toolInput;
  if (
    metadata &&
    typeof metadata === "object" &&
    Boolean((metadata as Record<string, unknown>).skip_persistence)
  ) {
    logDebug(hookName, "Skipping persistence (skip_persistence flag set)");
    return true;
  }
  return false;
}

/**
 * Emit hookSpecificOutput with additionalContext to stdout.
 * hookEventName is required by Claude Code's Zod validator (discriminated union).
 * Auto-detected from stdin payload (set by loadHookInput/runHook).
 *
 * SubagentStop and Stop events use top-level systemMessage field instead of hookSpecificOutput.
 * See SPEC.md §5.5
 */
export function emitContext(additionalContext: string): void {
  const eventName = _lastHookEvent ?? undefined;
  const tool = _lastToolName;

  // SubagentStop and Stop use top-level systemMessage field
  if (eventName === "SubagentStop" || eventName === "Stop") {
    const out = { systemMessage: additionalContext };
    process.stdout.write(JSON.stringify(out) + "\n");
    _logEmit("systemMessage", additionalContext.length, { event: eventName ?? "unknown", systemMessage: additionalContext });
    return;
  }

  // All other events use hookSpecificOutput
  const out: HookOutput = {
    hookSpecificOutput: {
      ...(eventName ? { hookEventName: eventName } : {}),
      additionalContext,
    },
  };
  const json = JSON.stringify(out);
  const eventDesc = tool ? `${eventName}:${tool}` : eventName ?? "unknown";
  _logEmit("context", additionalContext.length, { event: eventDesc, additionalContext });
  process.stdout.write(json + "\n");
}

/**
 * Emit hookSpecificOutput that denies the tool call with context and reason.
 * hookEventName is required by Claude Code's Zod validator (discriminated union).
 * Auto-detected from stdin payload (set by loadHookInput/runHook).
 * See SPEC.md §5.6
 */
export function emitContextAndBlock(
  additionalContext: string,
  reason: string,
): void {
  const eventName = _lastHookEvent ?? undefined;
  if (eventName && eventName !== "PreToolUse") {
    logWarn(_cachedHookName ?? "unknown",
      `emitContextAndBlock() called from ${eventName} — permissionDecision only works for PreToolUse. ` +
      `Use emitBlock() or the event-specific function instead.`);
  }
  const tool = _lastToolName;
  const out: HookOutput = {
    hookSpecificOutput: {
      ...(eventName ? { hookEventName: eventName } : {}),
      additionalContext,
      permissionDecision: "deny",
      permissionDecisionReason: reason,
    },
  };
  const json = JSON.stringify(out);
  const eventDesc = tool ? `${eventName}:${tool}` : eventName ?? "unknown";
  _logEmit("block", additionalContext.length, { event: eventDesc, additionalContext, blockReason: reason });
  process.stdout.write(json + "\n");
}

/** Log hook output (context, systemMessage, or block) to hook-log.jsonl for visibility. */
function _logEmit(type: "context" | "systemMessage" | "block", chars: number, payload: Record<string, unknown>): void {
  const hook = _cachedHookName ?? "unknown";
  const event = payload.event ?? "unknown";
  const mechanism = payload.mechanism ? ` via ${payload.mechanism}` : "";
  const msg = type === "block"
    ? `HOOK_OUTPUT [${type}] ${event} ${chars} chars${mechanism}, reason="${String(payload.blockReason ?? "").slice(0, 80)}"`
    : `HOOK_OUTPUT [${type}] ${event} ${chars} chars`;
  hookLog("info", hook, msg, { data: payload });
}

/**
 * Block a user prompt submission with a reason.
 * Only works for UserPromptSubmit hooks.
 * Output: top-level { decision: "block", reason } + optional hookSpecificOutput.additionalContext
 */
export function emitBlockPrompt(reason: string, context?: string): void {
  const eventName = _lastHookEvent ?? undefined;
  if (eventName && eventName !== "UserPromptSubmit") {
    logWarn(_cachedHookName ?? "unknown",
      `emitBlockPrompt() called from ${eventName} — only works for UserPromptSubmit`);
  }
  const out: HookOutput = {
    decision: "block",
    reason,
    ...(context ? {
      hookSpecificOutput: {
        ...(eventName ? { hookEventName: eventName } : {}),
        additionalContext: context,
      }
    } : {}),
  };
  _logEmit("block", context?.length ?? 0, { event: eventName ?? "unknown", additionalContext: context, blockReason: reason });
  process.stdout.write(JSON.stringify(out) + "\n");
}

/**
 * Block via exit code 2 + stderr feedback.
 * Works for PostToolUse, PostToolUseFailure.
 * The reason becomes the stderr message (fed to Claude as system-reminder).
 * If context is provided, it's prepended to the stderr message for richer feedback.
 * NOTE: Exit 2 causes Claude Code to ignore all JSON stdout — only stderr matters.
 */
export function emitBlockViaExit(reason: string, context?: string): void {
  const stderrMessage = context ? `${context}\n\n${reason}` : reason;
  _logEmit("block", stderrMessage.length, {
    event: _lastHookEvent ?? "unknown",
    blockReason: reason,
    mechanism: "exit2",
  });
  process.stderr.write(stderrMessage + "\n");
  throw new Error("SystemExit:2");
}

/**
 * Block via top-level { decision: "block", reason }.
 * Works for Stop and SubagentStop events.
 * These events do NOT support additionalContext — only reason is available.
 */
export function emitBlockTopLevel(reason: string): void {
  const eventName = _lastHookEvent ?? undefined;
  if (eventName && eventName !== "Stop" && eventName !== "SubagentStop") {
    logWarn(_cachedHookName ?? "unknown",
      `emitBlockTopLevel() called from ${eventName} — only works for Stop/SubagentStop`);
  }
  const out = { decision: "block", reason };
  _logEmit("block", reason.length, {
    event: eventName ?? "unknown",
    blockReason: reason,
    mechanism: "topLevelDecision",
  });
  process.stdout.write(JSON.stringify(out) + "\n");
}

/**
 * Respond to a PermissionRequest with allow/deny.
 * Only works for PermissionRequest hooks.
 */
export function emitPermissionDecision(
  behavior: "allow" | "deny",
  opts?: { message?: string; updatedInput?: Record<string, unknown>; updatedPermissions?: Record<string, unknown> },
): void {
  const out: PermissionRequestOutput = {
    decision: {
      behavior,
      ...(opts?.message ? { message: opts.message } : {}),
      ...(opts?.updatedInput ? { updatedInput: opts.updatedInput } : {}),
      ...(opts?.updatedPermissions ? { updatedPermissions: opts.updatedPermissions } : {}),
    },
  };
  _logEmit("block", 0, {
    event: _lastHookEvent ?? "unknown",
    blockReason: `permission:${behavior}`,
    mechanism: "permissionRequest",
  });
  process.stdout.write(JSON.stringify(out) + "\n");
}

/**
 * Unified block dispatcher — auto-detects the correct blocking mechanism
 * based on the current hook event type.
 *
 * PreToolUse → permissionDecision: "deny" (via emitContextAndBlock)
 * UserPromptSubmit → top-level decision: "block" (via emitBlockPrompt)
 * PostToolUse/PostToolUseFailure → exit(2) + stderr (via emitBlockViaExit)
 * Stop/SubagentStop → top-level { decision: "block", reason } (via emitBlockTopLevel)
 * PermissionRequest → decision: { behavior: "deny" } (via emitPermissionDecision)
 * SessionStart/Notification/SubagentStart/SessionEnd/etc. → warn and no-op
 *
 * This is the RECOMMENDED universal blocking API. Hook authors should use
 * emitBlock() and let the library handle event-specific dispatch.
 */
export function emitBlock(reason: string, context?: string): void {
  const event = _lastHookEvent;
  switch (event) {
    case "PermissionRequest": {
      emitPermissionDecision("deny", { message: reason });
      break;
    }
    case "PostToolUse":
    case "PostToolUseFailure": {
      emitBlockViaExit(reason, context);
      break;
    }
    case "PreToolUse": {
      emitContextAndBlock(context ?? reason, reason);
      break;
    }
    case "Stop":
    case "SubagentStop": {
      emitBlockTopLevel(reason);
      break;
    }
    case "UserPromptSubmit": {
      emitBlockPrompt(reason, context);
      break;
    }
    default: {
      logWarn(_cachedHookName ?? "unknown",
        `emitBlock() called from ${event ?? "unknown"} — no blocking mechanism exists for this event type, ignoring`);
      break;
    }
  }
}

/**
 * Auto-detect template origin from the hook script path.
 */
function detectTemplate(scriptPath = ""): string {
  const p = (scriptPath || (process.argv[1] ?? "")).replaceAll('\\', "/");
  if (
    p.includes("/_core/hooks-ts/") ||
    p.startsWith("_core/hooks-ts/") ||
    p.includes("/core/hooks-ts/") ||
    p.startsWith("core/hooks-ts/")
  ) {
    return "core";
  }
  const match = p.match(/_([a-z][a-z0-9-]*)\/hooks\//);
  if (match?.[1]) return match[1]; // e.g., "cc-native"
  return "unknown";
}

/**
 * Parse context window from hook input.
 * Returns [tokensUsed, maxTokens] or [null, null].
 * See SPEC.md §5.9
 */
export function parseContextWindow(
  hookInput: HookInput,
): [number | null, number | null] {
  const contextWindow = hookInput.context_window;
  if (!contextWindow) return [null, null];

  const currentUsage = contextWindow.current_usage;
  if (!currentUsage) return [null, null];

  const cacheRead = currentUsage.cache_read_input_tokens ?? 0;
  const inputTokens = currentUsage.input_tokens ?? 0;
  const cacheCreation = currentUsage.cache_creation_input_tokens ?? 0;
  const outputTokens = currentUsage.output_tokens ?? 0;

  const contentTokens = cacheRead + inputTokens + cacheCreation + outputTokens;
  const tokensUsed = contentTokens + CONTEXT_BASELINE_TOKENS;
  const maxTokens = contextWindow.context_window_size ?? DEFAULT_CONTEXT_WINDOW_SIZE;

  return [tokensUsed, maxTokens];
}

/**
 * Get context percentage remaining with fallback.
 * Returns [percentRemaining, tokensUsed, maxTokens] or [null, null, null].
 * See SPEC.md §5.9
 */
export function getContextPercentRemaining(
  hookInput: HookInput,
): [number | null, number | null, number | null] {
  const [tokensUsed, maxTokens] = parseContextWindow(hookInput);

  if (tokensUsed !== null && maxTokens !== null && maxTokens > 0) {
    const remaining = maxTokens - tokensUsed;
    const percentRemaining = Math.max(
      0,
      Math.min(100, Math.round((remaining / maxTokens) * 100)),
    );
    return [percentRemaining, tokensUsed, maxTokens];
  }

  // Source 2: context.json fallback (written by status_line.py)
  try {
    const sessionId = hookInput.session_id;
    if (sessionId) {
      const projectRoot = getProjectRoot(hookInput.cwd);
      const context = getContextBySessionId(sessionId, projectRoot);
      if (context?.last_session?.context_remaining_pct !== undefined) {
        return [context.last_session.context_remaining_pct, null, null];
      }
    }
  } catch {
    // Fallback failed — degrade gracefully
  }

  return [null, null, null];
}

/**
 * Read stdin early and extract session_id + event metadata.
 * Stashes parsed input for loadHookInput() to consume later.
 */
function _earlyReadInput(prefetchedInput?: Record<string, unknown>): void {
  if (prefetchedInput !== undefined) {
    _prefetchedInput = prefetchedInput;
  }

  // If we already have prefetched input, extract metadata from it
  if (_prefetchedInput && typeof _prefetchedInput === "object") {
    _lastHookEvent = readStringField(_prefetchedInput, "hook_event_name");
    _lastToolName = readStringField(_prefetchedInput, "tool_name");
    const sessionId = readStringField(_prefetchedInput, "session_id");
    if (sessionId) {
      setSessionId(sessionId);
    }
    return;
  }

  // Read stdin now so HOOK_START can include sid
  try {
    const inputData = fs.readFileSync(0, "utf8").trim();
    if (inputData) {
      const parsed = JSON.parse(inputData) as Record<string, unknown>;
      if (parsed && typeof parsed === "object") {
        _prefetchedInput = parsed;
        _lastHookEvent = readStringField(parsed, "hook_event_name");
        _lastToolName = readStringField(parsed, "tool_name");
        const sessionId = readStringField(parsed, "session_id");
        if (sessionId) {
          setSessionId(sessionId);
        }
      }
    }
  } catch {
    // Non-fatal — loadHookInput will return null
  }
}

/**
 * Standard hook entry point with lifecycle logging.
 * See SPEC.md §5.7
 */
export function runHook(
  mainFunc: () => number | void,
  hookName = "unknown",
  prefetchedInput?: Record<string, unknown>,
): never {
  _earlyReadInput(prefetchedInput);
  _cachedHookName = hookName;

  // Ensure cwd is project root so relative paths in hooks resolve correctly,
  // even when cwd has drifted via `cd` in a Bash tool call.
  try {
    const cwd = _prefetchedInput ? readStringField(_prefetchedInput, "cwd") ?? undefined : undefined;
    const projectRoot = getProjectRoot(cwd);
    if (process.cwd() !== projectRoot) process.chdir(projectRoot);
  } catch { /* non-fatal — proceed with current cwd */ }

  const startTime = performance.now();
  const template = detectTemplate();
  const event = _lastHookEvent ?? "unknown";
  const tool = _lastToolName;

  const startData: Record<string, unknown> = {
    lifecycle: "start",
    template,
    event,
  };
  if (tool) startData.tool = tool;
  hookLog("info", hookName, "HOOK_START", { data: startData });

  let exitCode = 0;
  let status = "success";
  let errorInfo: [Error, string] | null = null;

  try {
    const result = mainFunc();
    exitCode = typeof result === "number" ? result : 0;
    status = exitCode !== 0 ? "blocked" : "success";
  } catch (error: unknown) {
    if (error instanceof Error && error.message.startsWith("SystemExit:")) {
      const code = parseInt(error.message.slice(11), 10);
      exitCode = isNaN(code) ? (error.message.slice(11) ? 1 : 0) : code;
      status = exitCode !== 0 ? "blocked" : "success";
    } else {
      exitCode = 0; // Non-blocking
      status = "error";
      const stack = error instanceof Error ? error.stack ?? "" : "";
      errorInfo = [error instanceof Error ? error : new Error(String(error)), stack];
    }
  }

  _emitHookEnd(hookName, startTime, exitCode, status, errorInfo, startData, event, tool, template);
  process.exit(exitCode);
}

/**
 * Async variant of runHook for hooks that need await (e.g., AI inference).
 * Provides identical structured JSONL lifecycle logging as runHook.
 * See SPEC.md §5.7
 */
export function runHookAsync(
  mainFunc: () => Promise<number | void>,
  hookName = "unknown",
  prefetchedInput?: Record<string, unknown>,
): void {
  _earlyReadInput(prefetchedInput);
  _cachedHookName = hookName;

  // Ensure cwd is project root so relative paths in hooks resolve correctly,
  // even when cwd has drifted via `cd` in a Bash tool call.
  try {
    const cwd = _prefetchedInput ? readStringField(_prefetchedInput, "cwd") ?? undefined : undefined;
    const projectRoot = getProjectRoot(cwd);
    if (process.cwd() !== projectRoot) process.chdir(projectRoot);
  } catch { /* non-fatal — proceed with current cwd */ }

  const startTime = performance.now();
  const template = detectTemplate();
  const event = _lastHookEvent ?? "unknown";
  const tool = _lastToolName;

  const startData: Record<string, unknown> = {
    lifecycle: "start",
    template,
    event,
  };
  if (tool) startData.tool = tool;
  hookLog("info", hookName, "HOOK_START", { data: startData });

  mainFunc()
    .then((result) => {
      const exitCode = typeof result === "number" ? result : 0;
      _emitHookEnd(hookName, startTime, exitCode, exitCode !== 0 ? "blocked" : "success", null, startData, event, tool, template);
      _drainAndExit(exitCode);
    })
    .catch((error: unknown) => {
      let exitCode = 0;
      let status = "error";
      let errorInfo: [Error, string] | null = null;

      if (error instanceof Error && error.message.startsWith("SystemExit:")) {
        const code = parseInt(error.message.slice(11), 10);
        exitCode = isNaN(code) ? (error.message.slice(11) ? 1 : 0) : code;
        status = exitCode !== 0 ? "blocked" : "success";
      } else {
        exitCode = 0; // Non-blocking (fail open)
        const stack = error instanceof Error ? error.stack ?? "" : "";
        errorInfo = [error instanceof Error ? error : new Error(String(error)), stack];
      }

      _emitHookEnd(hookName, startTime, exitCode, status, errorInfo, startData, event, tool, template);
      _drainAndExit(exitCode);
    });
}

/** Shared HOOK_END logic for runHook and runHookAsync */
function _emitHookEnd(
  hookName: string,
  startTime: number,
  exitCode: number,
  status: string,
  errorInfo: [Error, string] | null,
  startData: Record<string, unknown>,
  event: string,
  tool: string | null,
  template: string,
): void {
  // Retroactive HOOK_START to per-context log (context_path resolved after main runs)
  const resolvedAfter = _getContextPath();
  if (resolvedAfter && fs.existsSync(resolvedAfter)) {
    hookLog("info", hookName, "HOOK_START", { data: startData });
  }

  const durationMs = Math.round((performance.now() - startTime) * 10) / 10;
  const endEvent = _lastHookEvent ?? event;
  const endTool = _lastToolName ?? tool;
  const endData: Record<string, unknown> = {
    lifecycle: "end",
    status,
    duration_ms: durationMs,
    exit_code: exitCode,
    template,
    event: endEvent,
  };
  if (endTool) endData.tool = endTool;

  if (errorInfo) {
    const [err, tb] = errorInfo;
    endData.error_type = err.constructor.name;
    hookLog("error", hookName, `[${endEvent}] ${err.constructor.name}: ${String(err).replaceAll(/[\n\r]/g, " ").slice(0, 200)}`, { traceback_str: tb });
    hookLog("error", hookName, `HOOK_END: ${err}`, { data: endData, traceback_str: tb });
  } else if (status === "blocked") {
    hookLog("warn", hookName, "HOOK_END", { data: endData });
  } else {
    hookLog("info", hookName, "HOOK_END", { data: endData });
  }
}

/**
 * Drain stdout before exiting to ensure pipe consumers receive all data.
 * On Windows, stdout to a pipe is fully buffered — process.exit() can
 * discard unflushed data. This waits for the write buffer to drain.
 */
function _drainAndExit(code: number): void {
  // If stdout is already finished or not writable, exit immediately
  if (!process.stdout.writable || process.stdout.writableFinished) {
    process.exit(code);
  }

  // Attempt to end stdout and wait for drain
  const timeout = setTimeout(() => process.exit(code), 1000); // safety fallback
  process.stdout.end(() => {
    clearTimeout(timeout);
    process.exit(code);
  });
}

export {hookLog, logBlocking, logDebug, logDiagnostic, logError, logHookError, logInfo, logWarn, setContextPath, setSessionId} from "../runtime/logger.js";

